src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Streaming.cs

```diff
@@ -0,0 +1,184 @@
+        {
+            // The framework is the only caller, so it's OK to have a nonobvious restriction like this
+            throw new InvalidOperationException($"{nameof(SendStreamingUpdatesAsync)} can only be called once.");
+        }
+
+        _streamingUpdatesWriter = writer;
+
+        try
+        {
+            await writer.FlushAsync(); // Make sure the initial HTML was sent
+            await untilTaskCompleted;
+        }
+        catch (NavigationException navigationException)
+        {
+            HandleNavigationAfterResponseStarted(writer, navigationException.Location);
+        }
+        catch (Exception ex)
+        {
+            HandleExceptionAfterResponseStarted(httpContext, writer, ex);
+
+            // The rest of the pipeline can treat this as a regular unhandled exception
+            // TODO: Is this really right? I think we'll terminate the response in an invalid way.
+            throw;
+        }
+    }
+
+    private void SendBatchAsStreamingUpdate(in RenderBatch renderBatch, TextWriter writer)
+    {
+        var count = renderBatch.UpdatedComponents.Count;
+        if (count > 0)
+        {
+            writer.Write("<blazor-ssr>");
+
+            // Each time we transmit the HTML for a component, we also transmit the HTML for its descendants.
+            // So, if we transmitted *every* component in the batch separately, there would be a lot of duplication.
+            // The subtrees projected from each component would overlap a lot.
+            //
+            // To avoid duplicated HTML transmission and unnecessary work on the client, we want to pick a subset
+            // of updated components such that, when we transmit that subset with their descendants, it includes
+            // every updated component without any duplication.
+            //
+            // This is quite easy if we first sort the list into depth order. As long as we process parents before
+            // their descendants, we can keep a log of the descendants we rendered, and then skip over those if we
+            // see them later in the list. This also implicitly handles the case where a batch contains the same
+            // root component multiple times (we only want to emit its HTML once).
+
+            // First, get a list of updated components in depth order
+            // We'll stackalloc a buffer if it's small, otherwise take a buffer on the heap
+            var bufSizeRequired = count * Marshal.SizeOf<ComponentIdAndDepth>();
+            var componentIdsInDepthOrder = bufSizeRequired < 1024
+                ? MemoryMarshal.Cast<byte, ComponentIdAndDepth>(stackalloc byte[bufSizeRequired])
+                : new ComponentIdAndDepth[count];
+            for (var i = 0; i < count; i++)
+            {
+                var componentId = renderBatch.UpdatedComponents.Array[i].ComponentId;
+                componentIdsInDepthOrder[i] = new(componentId, GetComponentDepth(componentId));
```

Comment:
Is it worth keeping the depth inside the `EndpointComponentState` to avoid this look-up?